### PR TITLE
Add delegation index to avoid looping through all delegations

### PR DESCRIPTION
**Pull request by rlan35**
Tested locally.

### CODE REVIEW

File: core/blockchain.go [hunk 1]
```diff
@@ -2555,6 +2540,35 @@ func (bc *BlockChain) UpdateStakingMetaData(tx *staking.StakingTransaction) erro
...
 }
 
+func (bc *BlockChain) addDelegationIndex(delegatorAddress, validatorAddress common.Address, root common.Hash) error {
+	delegations, err := bc.ReadDelegationsByDelegator(delegatorAddress)
+	if err != nil {
+		return err
+	}
+	found := false
+	for _, delegation := range delegations {
+		if bytes.Compare(delegation.ValidatorAddress.Bytes(), validatorAddress.Bytes()) == 0 {
```
Comment (fxfactorial) [inline core/blockchain.go]: suggest pull `validatorAddress.Bytes()` out of the loop to not byte thrash, copy memory over, over

File: core/state_transition.go [hunk 2]
```diff
@@ -496,31 +496,28 @@ func (st *StateTransition) applyCollectRewards(collectRewards *staking.CollectRe
...
-			if bytes.Equal(delegation.DelegatorAddress.Bytes(), collectRewards.DelegatorAddress.Bytes()) {
-				if delegation.Reward.Cmp(big.NewInt(0)) > 0 {
-					totalRewards.Add(totalRewards, delegation.Reward)
-				}
-
-				delegation.Reward.SetUint64(0)
-				break
+		if uint64(len(wrapper.Delegations)) > delegations[i].Index {
+			delegation := &wrapper.Delegations[delegations[i].Index]
+			if delegation.Reward.Cmp(big.NewInt(0)) > 0 {
```
Comment (fxfactorial) [inline core/state_transition.go]: `> 0` you mean intended to say == `1`?

File: core/blockchain.go (hunk 1 above)
Comment (rlan35) [reply core/blockchain.go]: Done.

File: core/state_transition.go (hunk 2 above)
Comment (rlan35) [reply core/state_transition.go]: Yes, checking Reward > 0

**Pull request merged**